src/wasm-traversal.h

```diff
@@ -157,185 +149,209 @@ struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
     self->visitMemory(&module->memory);
     self->visitModule(module);
   }
-};
 
-template<typename ParentType>
-struct ChildWalker : public WasmWalkerBase<ChildWalker<ParentType>> {
-  ParentType& parent;
+  // Walk implementation. We don't use recursion as ASTs may be highly
+  // nested.
 
-  ChildWalker(ParentType& parent) : parent(parent) {}
+  // Tasks receive the this pointer and a pointer to the pointer to operate on
+  typedef void (*TaskFunc)(SubType*, Expression**);
```

Comment:
True, and the replacing can "race". Perhaps it's best to document as the visit method not being allowed to replace. Another option is to get rid of the generic visit, but it is quite nice for e.g. the Metrics pass. But perhaps there could be a special class with only the generic visitor for that kind of thing? Not sure what's best.